src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -59,6 +61,9 @@ public MsQuicConnection(IPEndPoint localEndPoint, IPEndPoint remoteEndPoint, Int
             SetCallbackHandler();
             SetIdleTimeout(TimeSpan.FromSeconds(120));
             if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+
+            // Don't create security configuration on msquic thread, start in the background and await when accepting connection.
+            _securityConfigTask = Task.Run(() => SetSecurityConfigForConnection(cert, certFilePath, privateKeyFilePath));
```

Comment:
Why don't we want to run it on msquic thread?

Reply:
It throws an exception saying INVALID_STATE. @nibanks, you can’t create a security config on the callback for NEW_CONNECTION, correct?

Reply:
Rule 1, don't block the thread for any length of time. Besides that, you can do whatever you want. That being said creating the security config is synchronous in user mode, so I'd recommend not doing it on the thread. It could take a long time.

Reply:
Hrm, is it blocking on I/O? Ideally this would be fully asynchronous.

Reply:
It shouldn't be blocking on IO. However, I now realize the exception is being thrown from calling SetParam(SecurityConfig), rather than creating the security config.
```
System.Net.Quic.QuicException : Could not set security configuration. Error Code: INVALID_STATE
```

From what it seems like, you can't set a parameter for the security config until after the NEW_CONNECTION callback is done. @nibanks?

Reply:
If you're on the callback thread, you **must** return it back, not via a SetParam call. Otherwise, you call SetParam.

Reply:
So as long as you are off the callback thread, it's safe to call SetParam then. The only thing I'm worried about is a race between starting a Task to create the security config and when that completes when compared to when the msquic callback finishes for new connection.

I see the Task.Run to create the security config as the best options right now.

Reply:
As long as you call SetParam on a different thread, there is no race. Feel free to call SetParam before returning from the callback. MsQuic will not process the call until after the callback returns.